src/lib/crypto/pbkdf2.ts

```diff
@@ -0,0 +1,95 @@
+import { numberToBinUint32BE } from '../format/format.js';
+
+import type { HmacFunction} from './crypto.js';
+import { hmacSha256, hmacSha512 } from './crypto.js';
+
+export enum Pbkdf2Errors {
+  invalidIterationsError = 'Invalid PBKDF2 Parameters: Iterations must be a positive integer',
+  invalidDerivedKeyLengthError = 'Invalid PBKDF2 Parameters: Derived Key Length must be a positive integer',
+  invalidHmacLength = 'Invalid HMAC length: HMAC length must be a positive integer',
+}
+
+/**
+ * An object representing the parameters to use with PBKDF2 (Password-Based Key Derivation Function 2).
+ */
+export interface Pbkdf2Parameters {
+  /** The length of the derived key in bytes. */
+  derivedKeyLength: number;
+  password: Uint8Array;
+  iterations: number;
+  salt: Uint8Array;
+}
+
+export const instantiatePbkdf2Function =
+  (hmacFunction: HmacFunction, hmacByteLength: number) =>
+  (parameters: Pbkdf2Parameters) => {
+    /* eslint-disable functional/no-let, functional/no-loop-statement, functional/no-expression-statement, no-bitwise, no-plusplus */
+    const { password, salt, iterations, derivedKeyLength } = parameters;
+
+    if (!Number.isInteger(iterations) || iterations <= 0) {
+      return Pbkdf2Errors.invalidIterationsError;
+    }
+
+    if (!Number.isInteger(derivedKeyLength) || derivedKeyLength <= 0) {
+      return Pbkdf2Errors.invalidDerivedKeyLengthError;
+    }
+
+    if (!Number.isInteger(hmacByteLength) || hmacByteLength <= 0) {
+      return Pbkdf2Errors.invalidHmacLength;
+    }
+
+    const iterationCountByteLength = 4;
+
+    const derivedKey = new Uint8Array(derivedKeyLength);
+    const block = new Uint8Array(salt.length + iterationCountByteLength);
+    block.set(salt, 0);
+
+    let destPos = 0;
+    const length = Math.ceil(derivedKeyLength / hmacByteLength);
+
+    for (let i = 1; i <= length; i++) {
+      const iterationUint32BEEncoded = numberToBinUint32BE(i);
+      block.set(iterationUint32BEEncoded, salt.length);
+
+      const T = hmacFunction(password, block);
+      let U = T;
+
+      for (let j = 1; j < iterations; j++) {
+        U = hmacFunction(password, U);
+
+        for (let k = 0; k < hmacByteLength; k++) {
+          // @ts-expect-error-next-line
```

Comment:
@bitjson No idea how to get rid of this one.

Algo is almost verbatim https://github.com/browserify/pbkdf2/blob/master/lib/sync-browser.js with a few minor changes (e.g. Buffer to Uint8Array).

I'm wondering if Typescript is doing something funky because we're doing a bitwise op. If you have any ideas, please let me know.

Reply:
Ok, looks like the error is just that `U[k]` `is possibly 'undefined'.` as `k` isn't guaranteed to be in the range of Uint8Array `U`. We're sure in this case, so you can make that line:
```ts
T[k] ^= U[k]!; // eslint-disable-line @typescript-eslint/no-non-null-assertion
```
(Breaking rules is fine in sensible places, the linting just forces us to be explicit and deliberate about it. 👍)